logging/logging.go

```diff
@@ -72,25 +96,76 @@ func GetLogLevel() int {
 	return LogDiscard
 }
 
-// CustomizeLogging changes logging format
-func CustomizeLogging(loggingFormat string, serviceName string) {
-	log.SetOutput(os.Stderr)
-	log.SetFormatter(logFormatterFor(loggingFormat, serviceName))
+var (
+	errNoWriter      = errors.New("output writer is not specified")
+	errNoServiceName = errors.New("service name is not specified")
+	errNoFormat      = errors.New("log format is not specified")
+)
 
-	log.Debugf("Changed logging format to %s", loggingFormat)
+type CustomizeBuilder struct {
+	writer        io.Writer
+	serviceName   string
+	loggingFormat string
+	hooks         []FormatterHook
 }
 
-func logFormatterFor(loggingFormat string, serviceName string) log.Formatter {
-	loggingFormat = strings.ToLower(loggingFormat)
+func Customize() *CustomizeBuilder {
+	return &CustomizeBuilder{}
+}
 
-	if loggingFormat == "json" {
-		return JSONFormatter(log.Fields{FieldKeyProduct: serviceName})
+func (c *CustomizeBuilder) SetOutput(w io.Writer) *CustomizeBuilder {
+	c.writer = w
+	return c
+}
 
-	} else if loggingFormat == "cef" {
-		return CEFFormatter(log.Fields{FieldKeyProduct: serviceName})
+func (c *CustomizeBuilder) SetServiceName(serviceName string) *CustomizeBuilder {
+	c.serviceName = serviceName
+	return c
+}
+
+func (c *CustomizeBuilder) SetFormat(loggingFormat string) *CustomizeBuilder {
+	c.loggingFormat = loggingFormat
+	return c
+}
+
+func (c *CustomizeBuilder) SetHooks(hooks []FormatterHook) *CustomizeBuilder {
+	c.hooks = hooks
+	return c
+}
+
+func (c *CustomizeBuilder) Complete() error {
+	if c.writer == nil {
+		return errNoWriter
+	}
+	if c.serviceName == "" {
+		return errNoServiceName
 	}
+	if c.loggingFormat == "" {
+		return errNoFormat
+	}
+	/* We do not check hooks field, since it can be nil (standard log entry processing) */
+	log.SetOutput(c.writer)
+	log.SetFormatter(logFormatterFor(c.loggingFormat, c.serviceName, c.hooks))
+	log.Debugf("Changed logging format to %s", c.loggingFormat)
+	return nil
+}
 
-	return TextFormatter()
+// CustomizeLogging changes logging format
+func CustomizeLogging(loggingFormat string, serviceName string) {
+	log.SetOutput(os.Stderr)
```

Comment:
if we add builder, we should use it everywhere to avoid duplication code. Can we change it to `Customize().SetOutput(os.Stderr).SetFormat(loggingFormat).Complete()`? If we do so, we don't need to sync log message `Changed logging format to %s", loggingFormat` in the builder and in this function, sync way how should be defined formatter etc. Plus if we will add some more complicated logic in our builder, we should not to sync it to this function too, just update builder.

Reply:
Yes. Agree